graph/src/components/ethereum/adapter.rs

```diff
@@ -661,31 +710,129 @@ pub trait EthereumAdapter: Send + Sync + 'static {
+        // Scan the block range from triggers to find relevant blocks
+        if !log_filter.is_empty() {
+            trigger_futs.push(Box::new(
+                eth.blocks_with_logs(&logger, subgraph_metrics.clone(), from, to, log_filter)
+                    .map(|logs: Vec<Log>| logs.into_iter().map(EthereumTrigger::Log).collect()),
+            ))
+        }
+
+        if !call_filter.is_empty() {
+            trigger_futs.push(Box::new(
+                eth.blocks_with_calls(&logger, subgraph_metrics.clone(), from, to, call_filter)
+                    .map(EthereumTrigger::Call)
+                    .collect(),
+            ));
+        }
+
+        match block_filter.contract_addresses.len() {
+            0 => (),
+            _ => {
+                // To determine which blocks include a call to addresses
+                // in the block filter, transform the `block_filter` into
+                // a `call_filter` and run `blocks_with_calls`
+                let call_filter = EthereumCallFilter::from(block_filter);
+                trigger_futs.push(Box::new(
+                    eth.blocks_with_calls(&logger, subgraph_metrics.clone(), from, to, call_filter)
+                        .map(|call| {
+                            EthereumTrigger::Block(
+                                EthereumBlockPointer::from(&call),
+                                EthereumBlockTriggerType::WithCallTo(call.to),
+                            )
+                        })
+                        .collect(),
+                ));
+            }
+        }
+
+        let logger1 = logger.clone();
+        Box::new(
+            trigger_futs
+                .concat2()
+                .join(self.clone().block_hash_by_block_number(&logger, to))
+                .map(move |(triggers, to_hash)| {
+                    let mut block_hashes: HashSet<H256> =
+                        triggers.iter().map(EthereumTrigger::block_hash).collect();
+                    let mut triggers_by_block: HashMap<u64, Vec<EthereumTrigger>> =
+                        triggers.into_iter().fold(HashMap::new(), |mut map, t| {
+                            map.entry(t.block_number()).or_default().push(t);
+                            map
+                        });
+
+                    debug!(logger, "Found {} relevant block(s)", block_hashes.len());
+
+                    // Make sure `to` is included, even if empty.
+                    block_hashes.insert(to_hash.unwrap());
+                    triggers_by_block.entry(to).or_insert(Vec::new());
+
+                    (block_hashes, triggers_by_block)
+                })
+                .and_then(move |(block_hashes, mut triggers_by_block)| {
+                    self.load_blocks(logger1, chain_store, block_hashes)
+                        .map(move |block| {
+                            EthereumBlockWithTriggers::new(
+                                // All blocks with triggers should are in `triggers_by_block`, and
```

Comment:
Typo here: "... triggers should are in...".

Reply:
done thanks